go-controller/pkg/metrics/master.go

```diff
@@ -681,6 +713,374 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+	// takes to roll out the configuration to all nodes. To make matters more uncertain, the measurements are currently
+	// sampled.
+	cdr.reportKind[PodKindName] = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: MetricOvnkubeNamespace,
+		Subsystem: MetricOvnkubeSubsystemMaster,
+		Name:      "network_programming_pod_duration_seconds",
+		Help:      "Upperbound time to configure a pod from when its seen and config is applied to all nodes",
+		Buckets:   prometheus.ExponentialBuckets(.01, 3, 11),
+	})
+	cdr.reportKind[ServiceKindName] = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: MetricOvnkubeNamespace,
+		Subsystem: MetricOvnkubeSubsystemMaster,
+		Name:      "network_programming_service_duration_seconds",
+		Help:      "Upperbound time to configure a service from when its seen and config is applied to all nodes",
+		Buckets:   prometheus.ExponentialBuckets(.01, 3, 11),
+	})
+	configDurationRegOnce.Do(func() {
+		prometheus.MustRegister(cdr.reportKind[PodKindName])
+		prometheus.MustRegister(cdr.reportKind[ServiceKindName])
+		prometheus.MustRegister(metricConfigDurationStartCount)
+		prometheus.MustRegister(metricConfigDurationEndCount)
+		prometheus.MustRegister(metricConfigDurationOvnLatency)
+	})
+
+	// we currently do not clean the follow channels up upon exit
+	cdr.startRecordCh = make(chan startRecordEvent, startRecordChSize)
+	cdr.endRecordCh = make(chan endRecordEvent, endRecordChSize)
+	go cdr.processChannelEvents(stop)
+
+	// determine if we should be enabled or not during runtime
+	cdr.enabled = true
+	cdr.runEnable(kube, checkPeriod, nodeLimit, stop)
+
+	// record measurements are only available during periods of time. We use a ticker to set the availability of measurements.
+	// The ticker duration is continuously updated and is proportional to the node count.
+	// Determine the period (for the ticker) which record measurements are available. This can be updated during the runtime.
+	availableTicker := time.NewTicker(10 * time.Minute)
+	cdr.runRecordAvailableDuration(kube, availableTicker, delayPerNode, checkPeriod, stop)
+	// update the availability of measurements based on the duration determined above
+	cdr.runRecordAvailable(availableTicker, stop)
+
+	nbClient.Cache().AddEventHandler(&cache.EventHandlerFuncs{
+		UpdateFunc: func(table string, old model.Model, new model.Model) {
+			if table != nbGlobalTable {
+				return
+			}
+			// disable adding items to queue if functionality is disabled due to large node count
+			if !cdr.enabled {
+				return
+			}
+			oldRow := old.(*nbdb.NBGlobal)
+			newRow := new.(*nbdb.NBGlobal)
+
+			if oldRow.HvCfg != newRow.HvCfg && oldRow.HvCfgTimestamp != newRow.HvCfgTimestamp && newRow.HvCfgTimestamp > 0 {
+				select {
+				case cdr.endRecordCh <- endRecordEvent{hvCfg: newRow.HvCfg, hvCfgTimestamp: newRow.HvCfgTimestamp}:
+				default:
+					klog.Warning(chFullMsg)
```

Comment:
Should we delete corresponding `startRecord` here, because otherwise the next `hv_cfg` update will evaluate the wrong config duration for it?

Reply:
Nice spot. Will add update.